e2e: scanner.cy.ts and settings-shelters.cy.ts flake when the full local Cypress suite runs together

## Problem

Running the full local Cypress suite (\`npm run cy:run-ci-local\`, all 28 specs, no \`--spec\` filter)
intermittently fails two tests that are unrelated to each other and pass reliably when run in
isolation (\`npx cypress run --spec "cypress/e2e/<file>.cy.ts"\`):

1. \`scanner.cy.ts\` → \`Scanner\` → \`remains usable on mobile viewports\`
   ```
   HttpErrorResponse: ... Http failure response for http://localhost:4200/api/scanners/register: 403 Forbidden
   ```
   (an uncaught app error, so Cypress fails the test outright)

2. \`settings-shelters.cy.ts\` → \`Settings - Shelters\` → \`edits a shelter\`
   ```
   AssertionError: Timed out retrying after 10000ms: Expected to find element:
   `[testid^="searchresult-edituser-button-"]`, but never found it.
   ```

Both passed cleanly (3/3 and 8/8 respectively) when re-run individually right after the full-suite
failure, with no code changes in between - so this looks like state/timing that only shows up when
all specs share one backend/database instance back-to-back over a full ~5-6 minute run, not a
deterministic bug in either spec by itself. Root cause not yet diagnosed (could be scanner
registration id/session collision for #1, and load-induced timing for #2 - the shelters test's
selector itself is legitimate, just reused from `user-search.component.html`'s copy-pasted
`searchresult-edituser-button-` testid).

## How this was found

Found incidentally while verifying PR #2963 (customer-detail cost-contribution dropdown) didn't
break anything, by running the full local e2e suite as a smoke check. Neither failure has any
connection to the household/customer-detail code that PR touches - both are pre-existing and
reproduce the same way without that change.

## Suggested next step

Reproduce a few more full-suite runs to see how often each fails, then dig into whether it's
backend load, a shared-session issue, or spec ordering/isolation (e.g. does `scanner.register`
depend on state from an earlier spec's scanner).

--- a/frontend/src/main/webapp/src/app/common/http/xsrf-interceptor.service.spec.ts
+++ b/frontend/src/main/webapp/src/app/common/http/xsrf-interceptor.service.spec.ts
@@ -1,5 +1,5 @@
 import type { MockedObject } from 'vitest';
-import { HttpClient, provideHttpClient, withInterceptors, withXhr } from '@angular/common/http';
+import { HttpClient, HttpErrorResponse, provideHttpClient, withInterceptors, withXhr } from '@angular/common/http';
 import { TestBed } from '@angular/core/testing';
 import { HttpTestingController, provideHttpClientTesting } from '@angular/common/http/testing';
 import { xsrfInterceptor } from './xsrf-interceptor.service';
@@ -68,6 +68,36 @@ describe('XsrfInterceptor', () => {
 
         const request = httpTestingController.expectOne('/test');
         expect(request.request.headers.has('X-XSRF-TOKEN')).toBe(false);
+    });
+
+    it('retries once with a refreshed token after a 403 if the cookie has since appeared', () => {
+        cookieServiceSpy.get.mockReturnValueOnce('').mockReturnValueOnce('fresh-token');
+
+        let result: unknown;
+        httpClient.post('/test', {}).subscribe({error: (err) => result = err});
+
+        const firstRequest = httpTestingController.expectOne('/test');
+        expect(firstRequest.request.headers.has('X-XSRF-TOKEN')).toBe(false);
+        firstRequest.flush(null, {status: 403, statusText: 'Forbidden'});
+
+        const retriedRequest = httpTestingController.expectOne('/test');
+        expect(retriedRequest.request.headers.get('X-XSRF-TOKEN')).toBe('fresh-token');
+        retriedRequest.flush({});
+
+        expect(result).toBeUndefined();
+    });
+
+    it('does not retry a 403 when the cookie is unchanged', () => {
+        cookieServiceSpy.get.mockReturnValue('token-value');
+
+        let result: unknown;
+        httpClient.post('/test', {}).subscribe({error: (err) => result = err});
+
+        const request = httpTestingController.expectOne('/test');
+        request.flush(null, {status: 403, statusText: 'Forbidden'});
+
+        expect(result).toBeInstanceOf(HttpErrorResponse);
+        expect((result as HttpErrorResponse).status).toBe(403);
     });
 
 });

--- a/frontend/src/main/webapp/src/app/common/http/xsrf-interceptor.service.ts
+++ b/frontend/src/main/webapp/src/app/common/http/xsrf-interceptor.service.ts
@@ -1,6 +1,7 @@
-import {HttpEvent, HttpHandlerFn, HttpInterceptorFn, HttpRequest} from '@angular/common/http';
+import {HttpErrorResponse, HttpEvent, HttpHandlerFn, HttpInterceptorFn, HttpRequest} from '@angular/common/http';
 import {inject} from '@angular/core';
-import {Observable} from 'rxjs';
+import {Observable, throwError} from 'rxjs';
+import {catchError} from 'rxjs/operators';
 import {CookieService} from 'ngx-cookie-service';
 
 /**
@@ -18,10 +19,28 @@ export const xsrfInterceptor: HttpInterceptorFn = (
     return next(request);
   }
 
-  const token = cookieService.get('XSRF-TOKEN');
-  if (!token) {
-    return next(request);
-  }
+  const send = (token: string | undefined): Observable<HttpEvent<unknown>> =>
+    next(token ? request.clone({headers: request.headers.set('X-XSRF-TOKEN', token)}) : request);
 
-  return next(request.clone({headers: request.headers.set('X-XSRF-TOKEN', token)}));
+  const initialToken = cookieService.get('XSRF-TOKEN');
+
+  return send(initialToken).pipe(
+    catchError((error: unknown) => {
+      if (!(error instanceof HttpErrorResponse) || error.status !== 403) {
+        return throwError(() => error);
+      }
+
+      // The XSRF-TOKEN cookie is only set once some backend response has come back (it can't be
+      // primed by the initial page load, since that's served by the frontend dev server / static
+      // hosting, not the backend). A mutating request fired concurrently with the app's other
+      // bootstrap calls - e.g. the scanner page's registration call right after login - can race
+      // ahead of the response that would have set the cookie and go out with no token at all.
+      // Retry once with whatever the cookie is now before treating this as a genuine failure.
+      const freshToken = cookieService.get('XSRF-TOKEN');
+      if (!freshToken || freshToken === initialToken) {
+        return throwError(() => error);
+      }
+      return send(freshToken);
+    })
+  );
 };

